### PR TITLE
Redesign script config

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

- Redesign script configuration settings
- One of a key change is that script modifications are no longer "destructive" _until_ the settings are accepted via [OK] or [Apply] buttons. 
The race condition where a user has multiple instances of the app open and modifies scripts in several instances at once still remains, and it completely outside the scope of this PR. 
The last instance of the app to accept changes to scripts will overwrite all other changes may have been made in other instances. 

### Before

image

### After

image
image

## Test methodology 

- manual
- I may look at adding tests, but may do that at a later stage.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
Current limitations:
* No localisation support. The propertygrid categories and descriptors are compile-time constants. There's probably a way to change them at run-time, but I will need to investigate it.
* No image preview in the propertygrid. I'm sure it is possible, but further investigation is required.

The first limitation is probably a bigger one, and I'll try to sort it out for the v4 release. The second one I wouldn't consider a high priority or blocking.

File: GitUI/Design/ExecutableFileNameEditor.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Windows.Forms.Design;
+
+namespace GitUI.Design
+{
+    internal class ExecutableFileNameEditor : FileNameEditor
+    {
+        protected override void InitializeDialog(OpenFileDialog openFileDialog)
+        {
+            base.InitializeDialog(openFileDialog);
+            openFileDialog.Filter = "Executable file (*.exe;*.cmd;*.bat)|*.exe;*.cmd;*.bat|All files (*.*)|*.*";
```
Comment (mstv) [inline GitUI/Design/ExecutableFileNameEditor.cs:11]: + "*.ps1"?

**Comment by gerhardol** [comment]
> I'd like to see at least also Enabled, Event and perhaps Contexts in the list of scripts. => Vertical layout again?

Enabled could be seen in the list but more than that is harder (groups can handle one setting). 
This looks nicer (not like most GE forms...) but is less functional. This is no strong preference though
No comment on the code yet

**Comment by RussKie** [comment]
> Surprisingly, the script execution does not work anymore:

This was reported in gitter as well. Likely regressed in #7950, I'll have a look



> I'd like to see at least also `Enabled`, `Event` and perhaps `Contexts` in the list of scripts.

`Enabled` is expressed as color, enabled => black, disabled => grey. The same as in `FormRemotes`.
Showing `Event` is a good idea. 
`Context` - what benefit in that? Especially if there may be multiple contexts, e.g. revision grid, file tree, left panel, etc.?



> Vertical layout again?
> (If we had more lists than this and the Revision Links, we would also need a generalized list editor.)

Apologies, I'm struggling to parse this.



> (groups can handle one setting).

I had initially added groups, the same as we have in `FormRemotes`. However that made ordering scripts (i.e. up/down) significantly harder and mentally taxing. 
I think the order in scripts can be very important, e.g. a user may have several scripts that react to the same event, and they need to be executed in a set order.


> This looks nicer (not like most GE forms...)

Thank you, we need to start somewhere. I hope that eventually we can make the rest of the UI look at least as good.

> ...but is less functional.

What is in your opinion become less functional? And why is the old design more functional?

File: GitUI/Design/ExecutableFileNameEditor.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/Design/ExecutableFileNameEditor.cs:11]: I considered that, but PowerShell scripts are executed by PowerShell interpreter, that must be invoked explicitly. E.g. `> powershell.exe .\somescript.ps1`

**Comment by RussKie** [comment]
One more thing, we don't have any data<sup>1</sup>, but somehow I feel that scripts are very much underused. So the UI is built under an assumption that a majority of our users won't have more than a dozen of scripts.
 

--

1. except for what I observed myself and none of the users I observed was aware of scripts, let alone used them.

**Comment by vbjay** [comment]
I use them I have a couple I use all the time like stage non whitespace and
get current that does the heavy lifting of getting all submodules and the
main repo pulled down and then pushed to my forks.

**Comment by RussKie** [comment]
An alternative idea - scripts can be grouped by "Event", e.g.

```
--- AfterCommit ---
(icon) Fetch changes after commit

--- AfterPull ---
(icon) Update submodules after pull

--- ShowInUserMenuBar ---
(icon) Example

--- None ---
(icon) Open on GitHub
(icon) Fetch All Submodules
(icon) test
```

Obviously up/down ordering will be limited to the scope of a group, which probably makes sense.
Thoughts?

**Comment by RussKie** [comment]
image

**Comment by gerhardol** [comment]
I use one script all the time: submoudule update (sometimes I want to keep my checked out versions)
Other than that, I have found the triggers and available info to be a little too limited to use scripts a lot, so I use bash scripts.
There were some time since I looked in to this though.

The current view is functional, it shows a good overview in the grid. But it is not looking good, which is not important for first time users. 

With my usage, this change is not important, I can live with stepping through my scripts to see details, if Enabled is seen.
So @RussKie , I can give feedback but will probably not drive this change forward.

> An alternative idea - scripts can be grouped by "Event", e.g.

Other than Enable (Gray), this is the only dimension available. The trigger is probably not important enough to use that spot, it is more or less derived from the name anyway. Something like Show in Grid is dynamic and more important to me, but should not occupy a group.
Leave this unused for now.

**Comment by RussKie** [comment]
So you would prefer something like this? (more columns can be added)
image

**Comment by mstv** [comment]
> So would you prefer something like this? (more columns can be added)

Yep, this was what I meant with vertical layout.
But the current variant with `Enabled` -> Font color and column `Event` might suffice -- especially regarding the little usage.

Though I expect slightly increasing usage because we are going to add scripts to the context menus of files, folders and in the ROT.
Then I will not need to misuse the `Start shell` button in order to start VS Code.
And a column `Contexts` (plural) might help to get an overview and to find a specific script.

**Comment by RussKie** [comment]
I can make it vertical, no dramas.



> Though I expect slightly increasing usage because we are going to add scripts to the context menus of files, folders and in the ROT.
> Then I will not need to misuse the `Start shell` button in order to start VS Code.

That's the plan :)
I'm still trying to work out how to expose "selected file/folder" to a script.

**Comment by RussKie** [comment]
Btw, this is how you can configure VS Code:
image

**Comment by mstv** [comment]
> Btw, this is how you can configure VS Code:

Thank you! Kind of an easter egg, i.e. a hidden feature. This should be turned from `Event` to a `Script context`.

**Comment by vbjay** [comment]
image

**Comment by RussKie** [comment]
> Kind of an easter egg, i.e. a hidden feature. This should be turned from `Event` to a `Script context`.

Don't get me started on this... 😭 
I'm not going to touch it until after we release v4.

**Comment by pmiossec** [comment]
>Btw, this is how you can configure VS Code:

I just run the `code.cmd` (even if it require to specify the full path toward the file) and it works well.

**Comment by RussKie** [comment]
Updated the PR and the screenshot

**Comment by pmiossec** [comment]
Because space is limited in the grid, perhaps it could be interesting to merge the 'Command' and 'Arguments' in one column (and need to find a good name somewhat like 'full command').

**Comment by mstv** [comment]
Whishlist:
- [ ] Icon selector
- [ ] Splitter of the PropertyGrid should have a sensible position, not centered 
- [ ] column `Event` should fit to the longest event
- [ ] focus often jumps from PropertyGrid to the table above, e.g. try typing the name of an icon
- [ ] suppress bold font in PropertyGrid (I know that it indicates non-default values. But I see no benefit for our use case.)

**Comment by gerhardol** [comment]
This design is fine for me. The list has sufficient info (more columns could be added) and looks better.
I mostly agree with pmiossec and mstv's comments but that is a wishlist, so I could approve

Not adapted to dark theme: Icons need to be adapted and background set
image

**Comment by RussKie** [comment]
Thank you all for the suggestions



> perhaps it could be interesting to merge the 'Command' and 'Arguments' in one column

I suppose... Not finding it convincing, but I can do that.

> * Icon selector

This may have to wait (if we go with this design). My colleagues, who understand this functionality well, are very busy, and I can't seem to figure this one out in the time I have.


> * Splitter of the PropertyGrid should have a sensible position, not centered

I am not aware of how nor think `PropertyGrid` allows that changed. 
It is an [insanely complex control](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs), but it is short on customisation options. That was one of the reasons for the initial vertical design.



> * column `Event` should fit to the longest event

Agree, I will certainly fix this.

> * focus often jumps from PropertyGrid to the table above, e.g. try typing the name of an icon

I have missed that. I will fix this.

> * suppress bold font in PropertyGrid (I know that it indicates non-default values. But I see no benefit for our use case.)

I actually added "defaults" to un-bold. I find it easier to see which properties I have configured.
...and to make all the value normal weight it would require lots of extra code: https://stackoverflow.com/a/7442412/2338036




> Not adapted to dark theme: Icons need to be adapted and background set

Sure :) I just need to find how to do it, haha.

**Comment by RussKie** [comment]
> [x]  column `Event` should fit to the longest event
> [x] focus often jumps from PropertyGrid to the table above, e.g. try typing the name of an icon
> perhaps it could be interesting to merge the 'Command' and 'Arguments' in one column
> Not adapted to dark theme: Icons need to be adapted and background set

Addressed

**Comment by gerhardol** [comment]
@NikolayXHD Regarding dark mode in ScriptSettingsPage.cs OnPageShown()
 * Images are read from resources and Adapt is generally applied. Will the adaption have to be conditional to certain images or is it the images that need to be fixed?
 * The background for the control (System.Windows.Forms.ListView) is incorrect (AdaptBackColor() is not called, but no change if done). Any hints what need to be done?

Not critical for the PR itself

**Comment by RussKie** [comment]
Will hold off the merge until hear from @NikolayXHD

**Comment by NikolayXHD** [comment]
Fixed icon borders, property grid category header color.

Did not manage to fix checkboxes in ListView for now.

image

**Comment by RussKie** [comment]
Thank you all

**Comment by RussKie** [comment]
Added a wiki page: https://github.com/gitextensions/gitextensions/wiki/Useful-scripts

**Pull request merged**